### ISSUE TITLE
Low FPS During Video detection

### ISSUE THREAD

**Issue by jay0x5** [ultralytics/yolov5#6171]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Whenever I try to use my custom trained weights on a video not via the detect.py script but a custom inference script, I get very low frame rate like 1-2 per frame and video is very much sluggish
this is my script below as u can see i am trying to visualize the video using cv2.imshow method by opencv, but the frames are just very low like 1-2 fps
so I just wanted to ask if it affects the detection speed of model also or only the video looks sluggish to me but model is processing it normally?

Code:
```py
import yolov5
from vidgear.gears import CamGear
import cv2
import numpy as np

stream = CamGear(source='https://www.youtube.com/watch?v=N5VJWfXoDiM', stream_mode = True, logging=True).start() # YouTube Video URL as input
model = yolov5.load('C:\\Users\\Jay\\Desktop\\yolo-data\\YOLOV5\\yolov5-master\\weights\\best.pt')

# infinite loop

while True:
    
    frame = stream.read()

    # print(type(img))
    results = model(img)
    model.conf = 0.05
    results = model(img, size=1920)

    results = model(img, augment=True)
    label_id_offset = 1

    predictions = results.pred[0]
    boxes = predictions[:, :4] # x1, x2, y1, y2
    scores = predictions[:, 4].tolist()
    categories = predictions[:, 5].tolist()
    cv2.imshow("Detect", frame)

    print(scores)
    print(boxes)
    for category_id in categories:
        print(model.names[int(category_id)])


    key = cv2.waitKey(1) & 0xFF
    # check for 'q' key-press
    if key == ord("q"):
        #if 'q' key-pressed break out
        break

cv2.destroyAllWindows()
# close output window

# safely close video stream.
stream.stop()
    # print(results.pred[0].tolist())

**Comment by UygarUsta99** [comment]
Are you sure you are using gpu ? Can you try calling model.cuda() and then try again ? Also you are using size 1920  on inference which might slow down the inference time.

**Comment by jay0x5** [comment]
my cpu hits 100 whenever i run it
i have cuda and all installed but still my gpu is not accessed 
yes ill try reducing size
but is it my cpu causing the fps lag? and will fps lag affect my model's speed?

**Comment by SinDongHwan** [comment]
Hi, @jay0x5.
Try to check "torch.cuda.is_available()".
If return true, you installed torch with cuda.
but if return false, you installed torch without cuda.

if you want to run torch on gpu, 
Example) "pip install torch==1.8.1+cu111 torchvision==0.9.1+cu111 torchaudio==0.8.1 -f https://download.pytorch.org/whl/torch_stable.html".

You can refer install doc, [this link.](https://pytorch.org/get-started/previous-versions/)

**Comment by jay0x5** [comment]
Hi @SinDongHwan thanks for this but my question is will that fix my laggy fps?

**Comment by SinDongHwan** [comment]
Hi, @jay0x5.

When i see your situation, I think it uses cpu 100 hit because you didn't install torch with cuda.
So, i recommend to check "torch.cuda.is_available()".

**Comment by jay0x5** [comment]
Yes whenever i try to run it frames lag and cpu hits 100
Ill try the solution you said
Hopefully gpu doesnt make the frame lag

**Comment by jay0x5** [comment]
@SinDongHwan Hi i checked and my torch said it cannot access cuda
i have cuda 11.2 installed and cuDNN 10.2
and when i checked on website there is no torch installation command for cuda 11.2
should I go with cuda11.3? but again no cuDNN version for 11.3

**Comment by SinDongHwan** [comment]
Hi, @jay0x5.

I think you should select one of following versions.

`ERROR: Could not find a version that satisfies the requirement torch==1.8.1+cu112 (from versions: 1.7.1, 1.7.1+cpu, 1.7.1+cu101, 1.7.1+cu110, 1.8.0, 1.8.0+cpu, 1.8.0+cu101, 1.8.0+cu111, 1.8.1, 1.8.1+cpu, 1.8.1+cu101, 1.8.1+cu102, 1.8.1+cu111, 1.9.0, 1.9.0+cpu, 1.9.0+cu102, 1.9.0+cu111, 1.9.1, 1.9.1+cpu, 1.9.1+cu102, 1.9.1+cu111, 1.10.0, 1.10.0+cpu, 1.10.0+cu102, 1.10.0+cu111, 1.10.0+cu113, 1.10.1, 1.10.1+cpu, 1.10.1+cu102, 1.10.1+cu111, 1.10.1+cu113)`

if you use rtx 3000 serise graphic card, i recommend to install cuda 11.1 or 11.3.
if you want to use cuda 11.3, you can download "cuDNN 8.2.0 or 8.2.1 version for 11.x".

**Comment by jay0x5** [comment]
I am using a gtx1650 ill try one of those versions
thanks!

**Comment by wnfxsa1213** [comment]
> I am using a gtx1650 ill try one of those versions thanks!

 My graphics card is the same as yours.
I think your problem should be that your environment is not configured properly. 
My environment is CUDA 11.3 and cudnn 8.2.1

**Comment by jay0x5** [comment]
@wnfxsa1213 
my cuda is 11.2 and my cudnn is 10.2 so i think thats the issue

**Comment by anvinder** [comment]
If it helps anyone, i had the same problem on my 2070 super and updating to the following helps:
        - torch: 1.10.1+cu113
        - torchvision: 0.11
        - torchaudio: 0.10
        - CUDA version: 11.3
        - GPU driver version: 512.59
        - python version : 3.8
        - tensorflow:2.8.0
        - OS: Windows 10
        - command used:  pip install torch torchvision==0.11.2+cu113 torchaudio===0.10.1+cu113 -f https://download.pytorch.org/whl/cu113/torch_stable.html
NOTE: Please make sure you don't have a previous CUDA version on the same machine (initially i had both 11.6 and 11.3). If so please remove the  11.6 folders manually once you uninstall using control panel. 

Also, make sure to uninstall torch, torchvision and torchaudio using pip before running the above command. 
Hope it helps. Post this my machine can detect CUDA and the fps changed from 1 to 20.

**Comment by GitProgrammerGuy32** [comment]
Hey @jay0x5 buddy really wnated to know that if you are running yolov5 with this code then how am I supposed to run my yolov5 weights with the TFlite weights (I have converted already in an int8 format) is there a shorter code for that, because the code for TFlite I am using is quite huge and not working in the era of yolov8..

Thanks in advance
Best regards
Shaurya Grover

**Comment by glenn-jocher** [comment]
Hi @GitProgrammerGuy32,

To run YOLOv5 weights with TensorFlow Lite (TFLite), you can use the following steps:

1. Convert the YOLOv5 PyTorch weights to TensorFlow format using the `convert.py` script provided in the YOLOv5 repository. Make sure you have installed TensorFlow and ONNX dependencies.
2. Convert the PyTorch model to ONNX format using the `export.py` script from the YOLOv5 repository.
3. Convert the ONNX model to TensorFlow Lite format using the TensorFlow Lite Converter API or the `tflite_convert` command-line tool.

Here's a sample code snippet to demonstrate the process:

```python
import torch
import torch.nn as nn
import torchvision.models as models

# Load YOLOv5 PyTorch model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Export the model to ONNX format
torch.onnx.export(model, torch.randn(1, 3, 640, 640), 'yolov5s.onnx')

# Convert the ONNX model to TensorFlow Lite format
import tensorflow as tf
converter = tf.lite.TFLiteConverter.from_onnx_model('yolov5s.onnx')
tflite_model = converter.convert()

# Save the TFLite model to a file
with open('yolov5s.tflite', 'wb') as f:
    f.write(tflite_model)
```

Please note that conversion from YOLOv5 PyTorch weights to TFLite may have some differences in terms of performance and accuracy. You may need to experiment and fine-tune the converted model to achieve the desired results.

Hope this helps! Let me know if you have any further questions.

Best regards,
Glenn Jocher

**Comment by GitProgrammerGuy32** [comment]
Hi @glenn-jocher ,

Thanks for the explanation provided this is exactly what I needed but also wanted to understand how can I run the trained TFlite detections with the live webcam of my pc , is there a specific code provided for that which is short and effective and also wanted to know that rather than running yolov5 pre trained weights can I run my custom trained weights of yolov8? Thanks your for your assistance and interest in the issue.

Thanks in advance
Best regards
Shaurya grover

**Comment by glenn-jocher** [comment]
Hi @GitProgrammerGuy32,

To run the trained TFLite detections with the live webcam of your PC, you can use the OpenCV library in Python. Here's an example code snippet that demonstrates this:

```python
import cv2
import numpy as np
import tensorflow as tf

# Load the TFLite model
interpreter = tf.lite.Interpreter(model_path='yolov5s.tflite')
interpreter.allocate_tensors()

# Get the input and output details
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Open the webcam
cap = cv2.VideoCapture(0)

while True:
    # Read a frame from the webcam
    ret, frame = cap.read()

    # Perform any required pre-processing on the frame
    # ...

    # Set the input tensor
    input_data = np.expand_dims(frame, axis=0).astype(np.float32)
    interpreter.set_tensor(input_details[0]['index'], input_data)

    # Run the model
    interpreter.invoke()

    # Get the output tensor
    output_data = interpreter.get_tensor(output_details[0]['index'])

    # Perform any required post-processing on the output_data
    # ...

    # Visualize the results on the frame
    # ...

    # Display the frame
    cv2.imshow('Webcam', frame)

    # Exit if 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release the capture and close the windows
cap.release()
cv2.destroyAllWindows()
```

Regarding your question about running custom trained weights of YOLOv8, YOLOv8 is not officially available in the Ultralytics YOLOv5 repository. However, you can try to adapt the existing YOLOv5 codebase to incorporate your custom trained weights of YOLOv8. Keep in mind that this may require modifications to the model architecture and training pipeline.

I hope this answers your questions. Let me know if you need further assistance.

Thank you,
Glenn Jocher

**Comment by GitProgrammerGuy32** [comment]
> Hi @GitProgrammerGuy32,
> 
> To run YOLOv5 weights with TensorFlow Lite (TFLite), you can use the following steps:
> 
> 1. Convert the YOLOv5 PyTorch weights to TensorFlow format using the `convert.py` script provided in the YOLOv5 repository. Make sure you have installed TensorFlow and ONNX dependencies.
> 2. Convert the PyTorch model to ONNX format using the `export.py` script from the YOLOv5 repository.
> 3. Convert the ONNX model to TensorFlow Lite format using the TensorFlow Lite Converter API or the `tflite_convert` command-line tool.
> 
> Here's a sample code snippet to demonstrate the process:
> 
> ```python
> import torch
> import torch.nn as nn
> import torchvision.models as models
> 
> # Load YOLOv5 PyTorch model
> model = torch.hub.load('ultralytics/yolov5', 'yolov5s')
> 
> # Export the model to ONNX format
> torch.onnx.export(model, torch.randn(1, 3, 640, 640), 'yolov5s.onnx')
> 
> # Convert the ONNX model to TensorFlow Lite format
> import tensorflow as tf
> converter = tf.lite.TFLiteConverter.from_onnx_model('yolov5s.onnx')
> tflite_model = converter.convert()
> 
> # Save the TFLite model to a file
> with open('yolov5s.tflite', 'wb') as f:
>     f.write(tflite_model)
> ```
> 
> Please note that conversion from YOLOv5 PyTorch weights to TFLite may have some differences in terms of performance and accuracy. You may need to experiment and fine-tune the converted model to achieve the desired results.
> 
> Hope this helps! Let me know if you have any further questions.
> 
> Best regards, Glenn Jocher

Hey @glenn-jocher 
There are many errors I am facing in this code for my conversion.. so if possible can you provide me a collaboratory command 
from which I can simply convert my .pt trained weights (transferred learned from yolov5s.pt) to TFlite.

Thanks in Advance
Best regards
Shaurya grover

**Comment by glenn-jocher** [comment]
Hi @GitProgrammerGuy32,

To convert your .pt trained weights from YOLOv5 to TFLite using Google Colaboratory, you can follow these steps:

1. Upload your trained .pt weights file to Google Drive.
2. Open a new Python notebook in Google Colaboratory.
3. Mount your Google Drive by running the following code snippet:

```python
from google.colab import drive
drive.mount('/content/drive')
```

4. Install the required dependencies by running:

```python
!pip install torch torchvision
!pip install onnx
!pip install tfonnx
```

5. Convert the .pt weights to ONNX format by running the following code in the notebook:

```python
import torch

# Load the PyTorch model
model = torch.hub.load('ultralytics/yolov5', 'custom', path='/content/drive/your_model.pt')

# Export the model to ONNX format
torch.onnx.export(model, torch.randn(1, 3, 640, 640), '/content/drive/your_model.onnx')
```

Make sure to replace '/content/drive/your_model.pt' and '/content/drive/your_model.onnx' with the paths to your actual model files.

6. You can now follow the TensorFlow Lite conversion steps using the exported ONNX model as described in my previous response.

Remember to adjust the code snippets based on your specific file paths and requirements. Let me know if you encounter any further issues.

Cheers,
Glenn Jocher

**Comment by saleha-fatima** [comment]
I have read the whole discussion but my question persists, i have converted the yolov5s model to tflite model using this 

`git clone https://github.com/ultralytics/yolov5.git`
`!python yolov5/export.py --weights yolov5s.pt --include tflite`

then i tested this converted model on mobile and it generated 2FPS. I haven't done any finetuning atm. so i need to know what steps can i follow to increase the FPS during video detection.

**Comment by pderrenger** [comment]
To increase FPS during video detection with your TFLite model, consider optimizing the model by reducing the input image size, using a smaller model variant like YOLOv5n, or applying quantization techniques such as post-training quantization. Additionally, ensure your mobile device's hardware is capable of handling the model efficiently.

**Comment by saleha-fatima** [comment]
thanks for your response, I've tried reducing the input image size but first there was no apparent affect on FPS secondly it sometimes throw error saying that yolo only accepts 640 image size. As for the quantization, so while converting the yolov5 model to tflite it already quantized the model in fp16, And I've been doing inference with the quantized tflite model and there is no increase in fps. Regarding using the nano version, i will give it a try and let you know

**Comment by pderrenger** [comment]
You're welcome. If reducing the input size didn't help, ensure your model and input dimensions are consistent. Trying the YOLOv5n model is a good next step, as it's optimized for speed. Let us know how it goes.
